dial_sync.go

```diff
@@ -34,17 +41,61 @@ type activeDial struct {
 	id       peer.ID
 	refCnt   int
 	refCntLk sync.Mutex
+	ctx      context.Context
 	cancel   func()
 
+	addrs   map[string]struct{}
+	addrsLk sync.Mutex
+
 	err    error
 	conn   *Conn
 	waitch chan struct{}
+	connch chan *Conn
+	errch  chan error
+	dialch chan struct{}
+	donech chan struct{}
 
 	ds *DialSync
 }
 
-func (ad *activeDial) wait(ctx context.Context) (*Conn, error) {
+func (ad *activeDial) dial(ctx context.Context) (*Conn, error) {
 	defer ad.decref()
+
+	dialCtx := ad.dialContext(ctx)
```

Comment:
nit: by go convention, `xContext` means ["x but with a context"](https://golang.org/pkg/net/#Dialer.DialContext). Rename to `deriveDialContext`.